File: examples/examples-kotlin-java/src/test/java/com/mercadolibre/examples/java/UC04.java
package com.mercadolibre.examples.java;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.*;
import java.util.stream.Collectors;

class Collaborator04 {
	private String name;
	private String companyId;

	public String getName() {
		return name;
	}

	public String getCompanyId() {
		return companyId;
	}

	public Collaborator04(String name, String companyId) {
		this.name = name;
		this.companyId = companyId;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		Collaborator04 that = (Collaborator04) o;
		return Objects.equals(name, that.name) &&
				Objects.equals(companyId, that.companyId);
	}
}

class Office {
	private String name;
	private String companyId;

	public String getName() {
		return name;
	}

	public String getCompanyId() {
		return companyId;
	}

	public Office(String name, String companyId) {
		this.name = name;
		this.companyId = companyId;
	}

	@Override
	public String toString() {
		return "Office{" +
				"name='" + name + '\'' +
				", companyId='" + companyId + '\'' +
				'}';
	}
}

class Company {
	private String companyId;
	private List<String> collaborators;
	private List<String> offices;

	public Company(String companyId, List<String> collaborators, List<String> offices) {
		this.companyId = companyId;
		this.collaborators = collaborators;
		this.offices = offices;
	}

	public String getCompanyId() {
		return companyId;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		Company company = (Company) o;
		return Objects.equals(companyId, company.companyId) &&
				Objects.equals(collaborators, company.collaborators) &&
				Objects.equals(offices, company.offices);
	}
}

class InvalidCompanyException extends IllegalArgumentException {
	public InvalidCompanyException() {
		super("companyId is required");
	}
}

class OfficesService {
	public List<Company> compile(List<Collaborator04> collaborators, List<Office> offices) throws InvalidCompanyException {
		if (collaborators == null || offices == null) {
			return null;
		}
		validate(collaborators, offices);
		Map<String, List<Collaborator04>> collaboratorsCompanies = collaborators
				.stream()
				.collect(Collectors.groupingBy(Collaborator04::getCompanyId));
		Map<String, List<Office>> officesCompanies = offices
				.stream()
				.collect(Collectors.groupingBy(Office::getCompanyId));
		HashSet<String> companyIds = new HashSet<String>();
		companyIds.addAll(collaboratorsCompanies.keySet());
		companyIds.addAll(officesCompanies.keySet());
		return companyIds
					.stream()
					.map(it -> parseCompany(it, collaboratorsCompanies.get(it), officesCompanies.get(it)))
					.sorted(Comparator.comparing(Company::getCompanyId)).collect(Collectors.toList());
	}

	private Company parseCompany(String companyId, List<Collaborator04> collaborators, List<Office> offices) {
		return new Company(
			companyId,
			collaborators == null || collaborators.isEmpty() ? null : collaborators
				.stream()
				.map(Collaborator04::getName)
				.sorted()
				.collect(Collectors.toList()),
				offices == null || offices.isEmpty() ? null : offices
					.stream()
					.map(Office::getName)
					.sorted()
					.collect(Collectors.toList()));
	}

	private void validate(List<Collaborator04> collaborators, List<Office> offices) throws InvalidCompanyException {
		boolean collaboratorsAreValid = collaborators.stream().noneMatch(it -> it.getCompanyId() == null);
		boolean officesAreValid = offices.stream().noneMatch(it -> it.getCompanyId() == null);
		if (!collaboratorsAreValid || !officesAreValid) {
			throw new InvalidCompanyException();
		}
	}
}

public class UC04 {
	@Test
	public void UC04_01() throws InvalidCompanyException {
		List<Collaborator04> collaborators = Collections.singletonList(
			new Collaborator04("Maycow Antunes","Meli")
		);
		List<Office> offices = Collections.singletonList(new Office("Meli Floripa", "Meli"));
		OfficesService service = new OfficesService();
		List<Company> expected = List.of(
			new Company("Meli", List.of("Maycow Antunes"), List.of("Meli Floripa"))
		);
		Assertions.assertEquals(expected, service.compile(collaborators, offices));
	}

	@Test
	public void UC04_02() {
		List<Collaborator04> collaborators = Collections.singletonList(
			new Collaborator04("Maycow Antunes",null)
		);
		List<Office> offices = Collections.singletonList(new Office("Meli Floripa", "Meli"));
		OfficesService service = new OfficesService();
		Assertions.assertThrows(InvalidCompanyException.class, () -> service.compile(collaborators, offices));
	}

	@Test
	public void UC04_03() {
		List<Collaborator04> collaborators = Collections.singletonList(
				new Collaborator04("Maycow Antunes","Meli")
		);
		List<Office> offices = Collections.singletonList(new Office("Meli Floripa", null));
		OfficesService service = new OfficesService();
		Assertions.assertThrows(InvalidCompanyException.class, () -> service.compile(collaborators, offices));
	}

	@Test
	public void UC04_04() {
		List<Collaborator04> collaborators = Collections.singletonList(
			new Collaborator04("Maycow Antunes",null)
		);
		List<Office> offices = Collections.singletonList(new Office("Meli Floripa", null));
		OfficesService service = new OfficesService();
		Assertions.assertThrows(InvalidCompanyException.class, () -> service.compile(collaborators, offices));
	}

	@Test
	public void UC04_05() throws InvalidCompanyException {
		List<Collaborator04> collaborators = List.of(
			new Collaborator04("André Justi","Meli"),
			new Collaborator04("Wellington Macedo","Meli")
		);
		List<Office> offices = Collections.singletonList(new Office("Meli Floripa", "Meli"));
		OfficesService service = new OfficesService();
		List<Company> expected = List.of(
				new Company("Meli", List.of("André Justi", "Wellington Macedo"), List.of("Meli Floripa"))
		);
		Assertions.assertEquals(expected, service.compile(collaborators, offices));
	}

	@Test
	public void UC04_06() throws InvalidCompanyException {
		List<Collaborator04> collaborators = new ArrayList(Arrays.asList(
				new Collaborator04("Maycow Antunes","Meli"),
				new Collaborator04("Wellington Macedo","Meli"),
				new Collaborator04("Paulo Gustavo","Meli Envios")
			)
		);
		List<Office> offices = new ArrayList(Arrays.asList(
				new Office("Meli Floripa", "Meli"),
				new Office("Meli Cidade", "Meli")
			)
		);
		OfficesService service = new OfficesService();
		List<Company> expected = List.of(
			new Company(
				"Meli",
				List.of("Maycow Antunes", "Wellington Macedo"),
				List.of("Meli Cidade", "Meli Floripa")
			),
		new Company("Meli Envios", List.of("Paulo Gustavo"), null)
		);
		List<Company> c = service.compile(collaborators, offices);
		Assertions.assertEquals(expected, c);
	}

	@Test
	public void UC04_07() throws InvalidCompanyException {
		OfficesService service = new OfficesService();
		Assertions.assertNull(service.compile(null, null));
	}
}

File: examples/examples-kotlin-java/src/test/kotlin/com/mercadolibre/examples/kotlin/UC04.kt
package com.mercadolibre.examples.kotlin

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

data class Collaborator04(val name: String, val companyId: String?)
data class Office(val name: String, val companyId: String?)
data class Company(val companyId: String, val collaborators: List<String>?, val offices: List<String>?)

class InvalidCompanyException : IllegalArgumentException("companyId is required")


class CompanyService() {

	fun compile(collaborators: List<Collaborator04>, offices: List<Office>): List<Company> {
		validate(collaborators, offices)
		val collaboratorsCompanies = collaborators.groupBy(Collaborator04::companyId)
		val collaboratorsOffices = offices.groupBy(Office::companyId)
		return (collaboratorsCompanies.keys + collaboratorsOffices.keys)
				.distinct()
				.mapNotNull { parseCompany(it!!, collaboratorsCompanies[it], collaboratorsOffices[it]) }
				.sortedBy(Company::companyId)
	}

	private fun parseCompany(company: String, collaborators: List<Collaborator04>?, offices: List<Office>?) = Company(
			companyId = company,
			collaborators = collaborators?.map(Collaborator04::name)?.sorted(),
			offices = offices?.map(Office::name)?.sorted()
	)
}

private fun validate(collaborators: List<Collaborator04>, offices: List<Office>) {
	val collaboratorsAreValid = collaborators?.none { it.companyId == null }
	val officesAreValid = offices?.none { it.companyId == null }
	if (!collaboratorsAreValid || !officesAreValid) {
		throw InvalidCompanyException()
	}
}

class UC04 {

	private val companyService = CompanyService()

	@Test
	fun `UC04_01`() {
		val collaborators = listOf(Collaborator04("Maycow Antunes", "Meli"))
		val offices = listOf(Office("Meli Floripa", "Meli"))
		val companies = companyService.compile(collaborators, offices)
		val expected = listOf(
				Company("Meli", listOf("Maycow Antunes"), listOf("Meli Floripa"))
		)
		assertCompanies(expected, companies)
	}

	@Test
	fun `UC04_02`() {
		val collaborators = listOf(Collaborator04("Paulo Gustavo", null))
		val offices = listOf(Office("Meli Floripa", "Meli"))
		Assertions.assertThrows(InvalidCompanyException::class.java) {
			companyService.compile(collaborators, offices)
		}
	}

	@Test
	fun `UC04_03`() {
		val collaborators = listOf(Collaborator04("André Justi", null))
		val offices = listOf(Office("Meli Floripa", null))
		Assertions.assertThrows(InvalidCompanyException::class.java) {
			companyService.compile(collaborators, offices)
		}
	}

	@Test
	fun `UC04_04`() {
		val collaborators = listOf(Collaborator04("Paulo Gustavo", null))
		val offices = listOf(Office("Meli Floripa", null))
		Assertions.assertThrows(InvalidCompanyException::class.java) {
			companyService.compile(collaborators, offices)
		}
	}

	@Test
	fun `UC04_05`() {
		val collaborators = listOf(
				Collaborator04("André Justi", "Meli"),
				Collaborator04("Wellington Macedo", "Meli")
		)
		val offices = listOf(
				Office("Meli Floripa", "Meli")
		)
		val companies = companyService.compile(collaborators, offices)
		val expected = listOf(
				Company("Meli", listOf("André Justi", "Wellington Macedo"), listOf("Meli Floripa"))
		)
		assertCompanies(expected, companies)
	}

	@Test
	fun `UC04_06`() {
		val collaborators = listOf(
				Collaborator04("Maycow Antunes", "Meli"),
				Collaborator04("André Justi", "Meli"),
				Collaborator04("Paulo Gustavo", "Meli Envios")
		)
		val offices = listOf(
				Office("Meli Floripa", "Meli"),
				Office("Meli Cidade", "Meli")
		)
		val companies = companyService.compile(collaborators, offices)
		val expected = listOf(
				Company("Meli Envios", listOf("Paulo Gustavo"), null),
				Company("Meli", listOf("André Justi", "Maycow Antunes"), listOf("Meli Cidade", "Meli Floripa"))
		)
		assertCompanies(expected, companies)
	}

	@Test
	fun `UC04_07`() {
		val collaborators = emptyList<Collaborator04>()
		val offices = listOf<Office>()
		val companies = companyService.compile(collaborators, offices)
		assertCompanies(listOf(), companies)
	}

	private fun assertCompanies(expected: List<Company>, contains: List<Company>) {
		Assertions.assertTrue(expected.all { contains.contains(it) })
	}
}
